ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java

```diff
@@ -402,21 +405,26 @@ private String visitBooleanComparisonExpression(final ComparisonExpression.Type
     @Override
     protected Pair<String, Schema> visitCast(final Cast node, final Boolean context) {
       final Pair<String, Schema> expr = process(node.getExpression(), context);
-      final String returnTypeStr = node.getType();
-      final Schema returnType = SchemaUtil.getTypeSchema(returnTypeStr);
-      switch (returnTypeStr) {
+      final Type sqlType = node.getType();
+      if (!(sqlType instanceof PrimitiveType)) {
+        throw new KsqlFunctionException("Only casts to primitive types are supported: " + sqlType);
+      }
+
+      final Schema returnType = LogicalSchemas.fromSqlTypeConverter().fromSqlType(sqlType);
+      final Schema rightSchema = expr.getRight();
+      if (returnType.equals(rightSchema) || rightSchema == null) {
```

Comment:
This `if` statement is the actual fix for the `null` handling because the `rightSchema == null` results in the function returning a `{"null", returnType}` pair.  The `returnType.equals(rightSchema)` check also catches any no-op casts, i.e. casts from string to string, and allows the later logic to be simplified.